### PR TITLE
Global settings

### LINKED ISSUE #211: Global configuration
Hello,

It would be nice to be able to setup the plugin globally so that it is applied to all jobs at once, like global-slack-notifier plugin.

Thanks a lot

### PR DESCRIPTION
**Pull request by s-j-nilsson**
An attempt to enable global settings for the plugin. For now, the global settings are restricted to url and name. 

If no values for these properties are defined for an individual job, the values for the properties specified in the global settings are used.

Fixes #211

### CODE REVIEW

File: src/main/java/jenkins/plugins/office365connector/Webhook.java [hunk 1]
```diff
@@ -52,18 +54,23 @@
 
     private List<FactDefinition> factDefinitions = Collections.emptyList();
 
+    @Override
+    public DescriptorImpl getDescriptor() {
+            return (DescriptorImpl)super.getDescriptor();
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Webhook.java:59]: space is missing (formatting)

File: src/main/java/jenkins/plugins/office365connector/Webhook.java [hunk 2]
```diff
@@ -52,18 +54,23 @@
...
+    }
+
     @DataBoundConstructor
     public Webhook(String url) {
-        this.url = url;
+        this.url = StringUtils.isEmpty(url) ? getDescriptor().getUrl() : url;
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Webhook.java:64]: url seems to be mandatory - what does it change?

File: src/main/java/jenkins/plugins/office365connector/Webhook.java [hunk 3]
```diff
@@ -52,18 +54,23 @@
...
     }
 
     public String getUrl() {
         return url;
     }
 
     public String getName() {
-        // when the job is created and the name is not provided, then getName() returns null
-        return Util.fixEmptyAndTrim(name);
+        return Util.fixEmptyAndTrim(
```
Comment (damianszczepanik) [inline src/main/java/jenkins/plugins/office365connector/Webhook.java:72]: space at the end is not accepted?

File: src/test/java/jenkins/plugins/office365connector/WebhookTest.java [hunk 4]
```diff
@@ -28,6 +36,49 @@ public void getUrl_ReturnsUrl() {
...
+    public void getEmptyLocalUrl_ReturnsGlobalUrl() {
+
+        // given
+        String globalUrl = "globalUrl";
+        mockStatic(Jenkins.class);
+        Jenkins jenkins = mock(Jenkins.class);
+        Webhook.DescriptorImpl mockDescriptor = mock(Webhook.DescriptorImpl.class);
+
+        // when
+        when(Jenkins.getInstance()).thenReturn(jenkins);
```
Comment (damianszczepanik) [inline src/test/java/jenkins/plugins/office365connector/WebhookTest.java:49]: this is not "when" - this is still given

File: src/test/java/jenkins/plugins/office365connector/WebhookTest.java [hunk 5]
```diff
@@ -28,6 +36,49 @@ public void getUrl_ReturnsUrl() {
...
+
+        // then
+        Webhook webhook = new Webhook("");
+        String actualUrl = webhook.getUrl();
+
+        assertThat(actualUrl).isEqualTo(globalUrl);
+    }
+
+    @Test
+    public void getLocalUrl_ReturnsLocalUrlAndNotGlobal() {
```
Comment (damianszczepanik) [inline src/test/java/jenkins/plugins/office365connector/WebhookTest.java:61]: method mentioned in signature is not even invoked

File: src/test/java/jenkins/plugins/office365connector/WebhookTest.java [hunk 6]
```diff
@@ -28,6 +36,49 @@ public void getUrl_ReturnsUrl() {
...
+
+        // given
+        String globalUrl = "globalUrl";
+        String localUrl = "localUrl";
+        mockStatic(Jenkins.class);
+        Jenkins jenkins = mock(Jenkins.class);
+        Webhook.DescriptorImpl mockDescriptor = mock(Webhook.DescriptorImpl.class);
+
+        // when
```
Comment (damianszczepanik) [inline src/test/java/jenkins/plugins/office365connector/WebhookTest.java:70]: still //given

File: src/test/java/jenkins/plugins/office365connector/WebhookTest.java [hunk 7]
```diff
@@ -43,6 +94,50 @@ public void getName_ReturnsName() {
...
+        // given
+        String name = "myName";
+        Webhook webhook = new Webhook("someUrl");
+        webhook.setName(name);
+        mockStatic(Jenkins.class);
+        Jenkins jenkins = mock(Jenkins.class);
+        Webhook.DescriptorImpl mockDescriptor = mock(Webhook.DescriptorImpl.class);
+
+        // when
+        when(Jenkins.getInstance()).thenReturn(jenkins);
```
Comment (damianszczepanik) [inline src/test/java/jenkins/plugins/office365connector/WebhookTest.java:131]: same here and there - please review and fix

**Comment by damianszczepanik** [comment]
This is unclear to me: If no values for these properties are defined for an individual job, the values for the properties specified in the global settings are used.

Check why code coverage has decreased
Add description and screenshot to readme file

**Review by damianszczepanik — CHANGES_REQUESTED** [review]

File: src/main/java/jenkins/plugins/office365connector/Webhook.java (hunk 2 above)
Comment (s-j-nilsson) [reply src/main/java/jenkins/plugins/office365connector/Webhook.java:64]: It covers a use case we have where we in our Jenkinsfile add a webhook to a job where the URL is not specified. We want it to be picked up from the global settings that have already been saved.

`office365ConnectorWebhooks([[
            name: "MS Teams",
            notifySuccess: true,
            notifyFailure: true,
            notifyUnstable: true,
            notifyBackToNormal: true]])`

If we don't do the change in the constructor, the resulting job will not have a URL set for the webook.

File: src/main/java/jenkins/plugins/office365connector/Webhook.java (hunk 3 above)
Comment (damianszczepanik) [reply src/main/java/jenkins/plugins/office365connector/Webhook.java:72]: still many comments are not addressed

File: src/main/resources/jenkins/plugins/office365connector/Webhook/config.jelly [hunk 8]
```diff
@@ -2,10 +2,10 @@
 <j:jelly xmlns:j="jelly:core"
          xmlns:f="/lib/form">
     <f:entry title="URL" field="url">
-        <f:textbox/>
+        <f:textbox default="${descriptor.url}"/>
```
Comment (damianszczepanik) [inline src/main/resources/jenkins/plugins/office365connector/Webhook/config.jelly:5]: I have question/concern about this value. It is going to be saved with job configuration. Then then you update global value the one for build won't be updated.
Is it good, is it side effect ?

**Review by damianszczepanik — APPROVED** [review]
Let me know the answer for the last question so I will merge and test it a little bit before releasing

File: src/main/resources/jenkins/plugins/office365connector/Webhook/config.jelly (hunk 8 above)
Comment (s-j-nilsson) [reply src/main/resources/jenkins/plugins/office365connector/Webhook/config.jelly:5]: It is as we would like it to be. The global settings is meant to function as a default value when creating a job. This to facilitate for example, when you work with a Jenkinsfile and do not want to deal with specifying the URL but use whatever URL has been set globally for that Jenkins instance. I have updated this information in the README-file.

**Comment by damianszczepanik** [comment]
Sounds good, please resolve conflicts and let it go!

**Comment by samwar** [comment]
@s-j-nilsson Getting this merged would fix an incredible pain point in my jenkins deployment. Having to configure the connector for every job is a real PITA.

**Comment by s-j-nilsson** [comment]
Sorry, I have been indisposed due to Covid-19 but I have now rebased my branch and made the necessary adjustments, so hopefully everything looks ok now.

**Comment by s-j-nilsson** [comment]
I can see that Shippable is complaining about "Reversed (or previously applied) patch detected" but I can't spot what is wrong. From what I understand, it is related to an issue with how I rebased the code, but I can't see it. I would appreciate some assistance in resolving this issue.

**Comment by damianszczepanik** [comment]
@s-j-nilsson by default I see unhappy error at the global settings
obraz
can you fix it ?

**Comment by damianszczepanik** [comment]
I did some basic tests and looks good to me except this error even - please provide the fix so I can release it!

**Comment by s-j-nilsson** [comment]
Updated check on global settings page so that it does not show up by default.

**Comment by damianszczepanik** [comment]
Waiting for the PR with those changes...

**Pull request merged**